src/current/v24.1/certificate-based-authentication-using-the-x509-subject-field.md

```diff
@@ -0,0 +1,128 @@
+---
+title: Certificate-based authentication using multiple values from the X.509 Subject field
+summary: Learn about how to authenticate users using additional features already present in your X.509 security certificates
+toc: true
+docs_area: manage
+keywords: authentication, ldap, X.509, X509, tls
+---
+
+{% include_cached new-in.html version="v24.1" %} For customers that need to use their established Certificate Authority (CA) infrastructure to manage SQL user authentication, CockroachDB supports mapping of SQL user [roles]({% link {{ page.version.version }}/security-reference/authorization.md %}#roles) to values in the Subject field of the [X.509 certificate](https://en.wikipedia.org/wiki/X.509) used for TLS authentication.
+
+This mapping (hereafter referred to as "Subject mapping") can be used to automate the synchronization of SQL user roles with specific certificate attributes, leading to improved scalability of access control mechanisms.
+
+Subject mapping is useful if:
+
+- You run your own [Certificate Authority (CA)]({% link {{ page.version.version }}/security-reference/transport-layer-security.md %}#certificates-signing-trust-and-authority) infrastructure.
+- You need to use your existing CA infrastructure to manage [SQL user authentication]({% link {{ page.version.version }}/create-user.md %}#user-authentication).
+- You need to use the same CA for multiple CockroachDB clusters.
+
+For instructions showing how to map SQL user roles to values in the Subject field of the X.509 certificate, see the [Example](#example).
+
+{% include enterprise-feature.md %}
+
+## Related functionality
+
+Prior to CockroachDB v24.1, it was only safe to use certificates generated by CAs that placed restrictions on the `CommonName` field (so that only authorized users could get a cert with `CommonName=root`, for example). That meant that mapping the certificate principals to IP addresses, DNS names, and SQL users required using [the `--cert-principal-map` flag]({% link {{ page.version.version }}/cockroach-start.md %}#flags-cert-principal-map), which was [first added in v20.1]({% link releases/v20.1.md %}#v20-1-0-beta-3-changes-to-operational-workflows). That approach still works, but is superseded by the functionality described on this page, which should be much easier to use.
+
+The Subject mapping functionality described on this page makes it possible to use CAs that do not place restrictions on the `CommonName` field but do place restrictions on other subject fields (for example, that every cert has an `OU` field based on the user that created it).
+
+The [default workflow using `cockroach cert`]({% link {{ page.version.version }}/cockroach-cert.md %}#examples) only sets the `CommonName` field in its certificates, so the functionality described in this page is unnecessary. It is not possible to share a single CA managed by `cockroach cert` across multiple CockroachDB clusters.
```

Comment:
tl;dr:

I don't think the scope of this doc is going to allow for explaining that level of detail to the user here - we can say that the thing was not possible, but explaining why it was not possible is probably beyond the scope of what this doc is supposed to do 

my understanding is this doc is aimed at users who are doing fancy certificate things who won't need that level of background explanation, and/or who are just trying to get CA subject to user mapping done for their job 

@bdarnell please correct my understanding above if you disagree with my comments re: scope of this doc and whether we should talk about why `cockroach cert` works this way 

----

longer answer / research below:

as far as _why_ that is, I honestly don't know - that is how `cockroach cert` was designed and implemented a while back but for reasons that are beyond my knowledge (not a security expert)

I did look around and found some RFCs which talk about certs (rotation & revocation) but they don't really answer your question AFAICT

https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20170319_certificate_rotation.md
https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20200624_cert_revocation.md

I also did [this Glean search](https://app.glean.com/search?hcp=1&q=how+come+it%27s+not+possible+to+share+a+single+CA+managed+by+%60cockroach+cert%60+across+multiple+CockroachDB+clusters%3F) which did not answer the question but did turn up a bunch of cert and CA related links.  The Assistant basically said "I don't know"

> After reviewing the provided context, it appears that there is no specific document or snippet that directly explains why a single CA managed by cockroach cert cannot be shared across multiple CockroachDB clusters. The search results contain general information about the use of cockroach cert for creating certificates and keys, the use of custom CAs, and the importance of the CA in the trust store for node communication within a cluster. However, they do not address the limitations or reasons why a single CA cannot be used across multiple clusters.

> It's possible that the limitation is due to security best practices or architectural decisions within CockroachDB that ensure isolation and prevent cross-cluster certificate trust issues, but without specific documentation or explicit mention in the provided context, I cannot provide a definitive answer to your question.

> If you require a detailed explanation, I would recommend reaching out to the CockroachDB support team or consulting the official CockroachDB documentation for more information on this topic.

I also did a few minutes of grepping around in `~/go/src/github.com/cockroachdb/cockroach/pkg/security` but did not find an obvious answer within 5-10 minutes

Reply:
@bdarnell I also wanted to add that if you wouldn't mind answering 

> why It is not possible to share a single CA managed by cockroach cert across multiple CockroachDB clusters.

even if it's just a paragraph about the design, I am happy to make an internal wiki page with that info so it goes into Glean for our future search / reference as a company if this question ever comes up again

Reply:
You *can* use the same CA for multiple clusters, but you probably don't want to. The issue is that without subject mapping, all clusters that share a CA also share a user namespace. A `client.root.crt` for one cluster will be accepted as user `root` on *all* the clusters. Sometimes that may be what you want, as a crude kind of SSO, but in general if you're running multiple clusters you're going to want some sort of isolation between them instead of saying that access to one cluster implies access to all of them.

Reply:
thank you for that answer Ben!

I've made an [internal wiki page with that info](https://cockroachlabs.atlassian.net/wiki/x/P4BF1) and linked it back here for future reference

Reply:
so in this case it sounds like "it is not possible to use the same CA" means "it is not possible to use the same CA in a way that makes much sense from a security POV".  Rather than add another paragraph with the explanation Ben provided, I'd rather leave this text alone since it has already had SME review and been explained here a bit.  Also adding that add'l text would not really provide much value to users who are doing this subject mapping stuff AFAICT.  @florence-crl I hope that you feel like your question was answered - gonna resolve this thread